### PR TITLE
Added Floyd's Algorithm for Linked List Cycle Detection

### PR DESCRIPTION
**Pull request by Gyan172004**
Description:
I've added Floyd's Tortoise and Hare Algorithm for detecting cycles in linked lists.

Changes Made:
- Implemented the algorithm.
- Tested with sample cases.

### CODE REVIEW

**Review by siriak — CHANGES_REQUESTED** [review]
Please check file extensions
image

**Review by siriak — CHANGES_REQUESTED** [review]
This is a duplicate
image

**Comment by Gyan172004** [comment]
I have deleted my duplicate file which was not having file extension.
Please review.

**Review by siriak — CHANGES_REQUESTED** [review]
Please rename the file to `floyds_algorithm.rs` and use the existing implementation of linked list from this repo

**Review by siriak — COMMENTED** [review]
Could you use this implementation of LinkedList? https://github.com/TheAlgorithms/Rust/blob/9d57491f3b9527f5ef41c1a3f738b829653734da/src/data_structures/linked_list.rs

**Comment by Gyan172004** [comment]
I am facing some issues in using the existing implementation of linked list from this repo. in my code.
Should I use `std::collections::linked_list` , I think its better to use standard library inbuilt implementation?

**Comment by siriak** [comment]
Using our list would be ideal. Could you fix the problem that prevents you from using it? If you don't have time for that, you can just use `std::collections::linked_list`

**Comment by Gyan172004** [comment]
```rs
// floyds_algorithm.rs
// https://github.com/rust-lang/rust/blob/master/library/alloc/src/collections/linked_list.rs#L113
// use std::collections::linked_list::LinkedList;
// https://www.reddit.com/r/rust/comments/t7wquc/is_it_possible_to_solve_leetcode_problem141/


use crate::data_structures::linked_list::LinkedList; // Import the LinkedList from linked_list.rs

pub fn has_cycle<T>(linked_list: &LinkedList<T>) -> bool {
    let mut slow = linked_list.head;
    let mut fast = linked_list.head;

    let mut flag = false;

    while let (Some(slow_node), Some(fast_node)) = (slow, fast) {
        unsafe {
            slow = slow_node.as_ref().next;
            fast = fast_node.as_ref().next;

            if let Some(fast_next) = fast {
                // fast = (*fast_next.as_ptr()).next;
                fast = fast_next.as_ref().next;
            } else {
                return false; // If fast reaches the end, there's no cycle
            }

            if slow == fast {
                
                return true; // Cycle detected
            }
        }
        
    }
    // println!("{}", flag);
    false // No cycle detected
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_cycle_no_cycle() {
        let mut linked_list = LinkedList::new();
        linked_list.insert_at_tail(1);
        linked_list.insert_at_tail(2);
        linked_list.insert_at_tail(3);

        // assert_eq!(has_cycle(&linked_list), false);

        assert_eq!(detect_cycle(&mut linked_list), None);
    }

    #[test]
    fn test_detect_cycle_with_cycle() {
        let mut linked_list = LinkedList::new();
        linked_list.insert_at_tail(1);
        linked_list.insert_at_tail(2);
        linked_list.insert_at_tail(3);

        // Create a cycle for testing
        unsafe {
            if let Some(mut tail) = linked_list.tail {
                if let Some(head) = linked_list.head {
                    tail.as_mut().next = Some(head);
                }
            }
        }

        // assert_eq!(has_cycle(&linked_list), true);
        assert_eq!(detect_cycle(&mut linked_list), Some(3));
    }
}

```

I tried implementing one, but I have got a roadblock actually! On function return on true, SIGSEV occurs due to invalid memory reference, my initial presumption says, that it is due to improper Drop Traits implemented for LinkedList (but, I maybe wrong too). I have tried debugging this beyond a period that, I am willing to give up this approach , maybe for better (sorry!)

Hence, I believe std::collections::linked_list might be more easier for me to implement. However, if you can help me with debugging or recommending some more better approaches then also, I am more than welcome to do that

**Comment by ethanwater** [comment]
It the issue is connected to the linked_list you're using as you presumed:

```
running 2 tests
test data_structures::floyds_algorithm::tests::test_detect_cycle_no_cycle ... ok
thread 'data_structures::floyds_algorithm::tests::test_detect_cycle_with_cycle' panicked at 'attempt to subtract with overflow', src/data_structures/linked_list.rs:121:13
```


As the panic log states above it is due to this line of code here:` self.length -= 1;` in linked_list.rs:121:13. Your code is trying to access an address that doesn't exist as you probably already know. This could be due to the management of memory in your code, the infrastructure of the linked list you're using, or both. I also tried to look at it and debug but I got hungry and admittedly a bit frustrated. Below is some interesting data for you to analyze if you choose to commit towards debugging the issue.

```
; Symbol: rust_panic
; Source: /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:740
1006AAADC: FF 83 02 D1                sub    sp, sp, #0xa0
1006AAAE0: FD 7B 09 A9                stp    x29, x30, [sp, #0x90]
1006AAAE4: FD 43 02 91                add    x29, sp, #0x90
1006AAAE8: 9C 24 00 94                bl     0x1006b3d58  ; __rust_start_panic at lib.rs:102
1006AAAEC: E0 07 00 B9                str    w0, [sp, #0x4]
1006AAAF0: E8 13 00 91                add    x8, sp, #0x4
1006AAAF4: C9 00 10 10                adr    x9, #131096  ; core::fmt::num::imp::<impl core::fmt::Display for u32>::fmt at num.rs:279
1006AAAF8: 1F 20 03 D5                nop    
1006AAAFC: A8 A7 3E A9                stp    x8, x9, [x29, #-0x18]
1006AAB00: 48 1A 00 D0                adrp   x8, 842
1006AAB04: 08 21 2C 91                add    x8, x8, #0xb08  ; log::LOG_LEVEL_NAMES + 16968
1006AAB08: 29 00 80 52                mov    w9, #0x1  ; =1 
1006AAB0C: E8 A7 05 A9                stp    x8, x9, [sp, #0x58]
1006AAB10: FF 27 00 F9                str    xzr, [sp, #0x48]
1006AAB14: A8 63 00 D1                sub    x8, x29, #0x18
1006AAB18: E8 A7 06 A9                stp    x8, x9, [sp, #0x68]
1006AAB1C: 88 6F 0C 10                adr    x8, #101872  ; <core::fmt::Arguments as core::fmt::Display>::fmt at mod.rs:580
1006AAB20: 1F 20 03 D5                nop    
1006AAB24: EA 23 01 91                add    x10, sp, #0x48
1006AAB28: EA A3 03 A9                stp    x10, x8, [sp, #0x38]
1006AAB2C: 48 1A 00 B0                adrp   x8, 841
1006AAB30: 08 41 3D 91                add    x8, x8, #0xf50  ; log::LOG_LEVEL_NAMES + 13968
1006AAB34: 4A 00 80 52                mov    w10, #0x2  ; =2 
1006AAB38: E8 AB 01 A9                stp    x8, x10, [sp, #0x18]
1006AAB3C: FF 07 00 F9                str    xzr, [sp, #0x8]
1006AAB40: E8 E3 00 91                add    x8, sp, #0x38
1006AAB44: E8 A7 02 A9                stp    x8, x9, [sp, #0x28]
1006AAB48: A0 23 00 D1                sub    x0, x29, #0x8
1006AAB4C: E1 23 00 91                add    x1, sp, #0x8
1006AAB50: 0E ED FF 97                bl     0x1006a5f88  ; std::io::Write::write_fmt at mod.rs:1677
1006AAB54: CE 89 FF 97                bl     0x10068d28c  ; core::ptr::drop_in_place<core::result::Result<(),std::io::error::Error>> at mod.rs:490
1006AAB58: CA 0E 00 94                bl     0x1006ae680  ; std::sys::unix::abort_internal at mod.rs:349
```

**Review by ethanwater — COMMENTED** [review]
submitted a response on the SIGSEGV here: https://github.com/TheAlgorithms/Rust/pull/535#issuecomment-1756629169

**Comment by Gyan172004** [comment]
> As the panic log states above it is due to this line of code here: self.length -= 1; in linked_list.rs:121:13. Your code is trying to access an address that doesn't exist as you probably already know. This could be due to the management of memory in your code, the infrastructure of the linked list you're using, or both. I also tried to look at it and debug but I got hungry and admittedly a bit frustrated. Below is some interesting data for you to analyze if you choose to commit towards debugging the issue.

Thanks for the review! I just checked in and the found that Drop trait for LinkedList wasn't implemented properly for cyclic list, actually it linearly traverse and delete element one by one till there is a deadend, but for cyclic list there is no deadend, so this approach will fail and hence those **invalid memory reference**errors.

I have just pushed in the patch for LinkedList, can you check it out please?

File: src/data_structures/linked_list.rs [hunk 1]
```diff
@@ -111,16 +111,22 @@ impl<T> LinkedList<T> {
     pub fn delete_head(&mut self) -> Option<T> {
         // Safety: head_ptr points to a leaked boxed node managed by this list
         // We reassign pointers that pointed to the head node
+        if self.length == 0 {
+            return None;
+        }
+
         self.head.map(|head_ptr| unsafe {
+            println!("{}", self.length);
```
Comment (siriak) [inline src/data_structures/linked_list.rs:119]: No need to print logs

**Review by siriak — CHANGES_REQUESTED** [review]
Looks good, just please remove logging

**Review by siriak — APPROVED** [review]
Looks good, thanks!

**Comment by Gyan172004** [comment]
> It the issue is connected to the linked_list you're using as you presumed:
> 
> ```
> running 2 tests
> test data_structures::floyds_algorithm::tests::test_detect_cycle_no_cycle ... ok
> thread 'data_structures::floyds_algorithm::tests::test_detect_cycle_with_cycle' panicked at 'attempt to subtract with overflow', src/data_structures/linked_list.rs:121:13
> ```
> 
> As the panic log states above it is due to this line of code here:` self.length -= 1;` in linked_list.rs:121:13. Your code is trying to access an address that doesn't exist as you probably already know. This could be due to the management of memory in your code, the infrastructure of the linked list you're using, or both. I also tried to look at it and debug but I got hungry and admittedly a bit frustrated. Below is some interesting data for you to analyze if you choose to commit towards debugging the issue.
> 
> ```
> ; Symbol: rust_panic
> ; Source: /rustc/90c541806f23a127002de5b4038be731ba1458ca/library/std/src/panicking.rs:740
> 1006AAADC: FF 83 02 D1                sub    sp, sp, #0xa0
> 1006AAAE0: FD 7B 09 A9                stp    x29, x30, [sp, #0x90]
> 1006AAAE4: FD 43 02 91                add    x29, sp, #0x90
> 1006AAAE8: 9C 24 00 94                bl     0x1006b3d58  ; __rust_start_panic at lib.rs:102
> 1006AAAEC: E0 07 00 B9                str    w0, [sp, #0x4]
> 1006AAAF0: E8 13 00 91                add    x8, sp, #0x4
> 1006AAAF4: C9 00 10 10                adr    x9, #131096  ; core::fmt::num::imp::<impl core::fmt::Display for u32>::fmt at num.rs:279
> 1006AAAF8: 1F 20 03 D5                nop    
> 1006AAAFC: A8 A7 3E A9                stp    x8, x9, [x29, #-0x18]
> 1006AAB00: 48 1A 00 D0                adrp   x8, 842
> 1006AAB04: 08 21 2C 91                add    x8, x8, #0xb08  ; log::LOG_LEVEL_NAMES + 16968
> 1006AAB08: 29 00 80 52                mov    w9, #0x1  ; =1 
> 1006AAB0C: E8 A7 05 A9                stp    x8, x9, [sp, #0x58]
> 1006AAB10: FF 27 00 F9                str    xzr, [sp, #0x48]
> 1006AAB14: A8 63 00 D1                sub    x8, x29, #0x18
> 1006AAB18: E8 A7 06 A9                stp    x8, x9, [sp, #0x68]
> 1006AAB1C: 88 6F 0C 10                adr    x8, #101872  ; <core::fmt::Arguments as core::fmt::Display>::fmt at mod.rs:580
> 1006AAB20: 1F 20 03 D5                nop    
> 1006AAB24: EA 23 01 91                add    x10, sp, #0x48
> 1006AAB28: EA A3 03 A9                stp    x10, x8, [sp, #0x38]
> 1006AAB2C: 48 1A 00 B0                adrp   x8, 841
> 1006AAB30: 08 41 3D 91                add    x8, x8, #0xf50  ; log::LOG_LEVEL_NAMES + 13968
> 1006AAB34: 4A 00 80 52                mov    w10, #0x2  ; =2 
> 1006AAB38: E8 AB 01 A9                stp    x8, x10, [sp, #0x18]
> 1006AAB3C: FF 07 00 F9                str    xzr, [sp, #0x8]
> 1006AAB40: E8 E3 00 91                add    x8, sp, #0x38
> 1006AAB44: E8 A7 02 A9                stp    x8, x9, [sp, #0x28]
> 1006AAB48: A0 23 00 D1                sub    x0, x29, #0x8
> 1006AAB4C: E1 23 00 91                add    x1, sp, #0x8
> 1006AAB50: 0E ED FF 97                bl     0x1006a5f88  ; std::io::Write::write_fmt at mod.rs:1677
> 1006AAB54: CE 89 FF 97                bl     0x10068d28c  ; core::ptr::drop_in_place<core::result::Result<(),std::io::error::Error>> at mod.rs:490
> 1006AAB58: CA 0E 00 94                bl     0x1006ae680  ; std::sys::unix::abort_internal at mod.rs:349
> ```

I'm eager to learn which debugger tools you used to generate these logs
 it would be very helpful. Thanks!

**Pull request merged**